avm/src/daemon/ctrl.rs

```diff
@@ -0,0 +1,200 @@
+
+use crate::daemon::daemon::CLUSTER_SECRET;
+use crate::make_server;
+use crate::vm::http::{HttpType, HttpsConfig};
+
+#[derive(Debug)]
+pub struct HashedId {
+  pub id: String,
+  pub hash: u64,
+  pub is_up: bool,
+}
+
+impl HashedId {
+  pub fn new(id: String) -> HashedId {
+    let mut hasher = XxHash64::with_seed(0xa1a2);
+    hasher.write(id.as_bytes());
+    HashedId {
+      id,
+      hash: hasher.finish(),
+      is_up: true, // Assume it's up and let the cron job mutate it over time
+    }
+  }
+}
+
+// An algorithm based on the classic Rendezvous Hash but with changes to make the performance
+// closer to modern Consistent Hash algorithms while retaining Rendezvous Hash's coordination-free
+// routing.
+#[derive(Debug)]
+pub struct LogRendezvousHash {
+  sorted_hashes: Vec<HashedId>,
+}
+
+impl LogRendezvousHash {
+  pub fn new(ids: Vec<String>) -> LogRendezvousHash {
+    let mut sorted_hashes = Vec::with_capacity(ids.len());
+    for id in ids {
+      sorted_hashes.push(HashedId::new(id));
+    }
+    sorted_hashes.sort_by(|a, b| a.hash.cmp(&b.hash));
+    LogRendezvousHash { sorted_hashes }
+  }
+
+  /// Replaces the prior set of ids with the new set. TODO: Skip re-hashing ids that already exist
+  pub fn update(self: &mut LogRendezvousHash, ids: Vec<String>) {
+    let mut sorted_hashes = Vec::with_capacity(ids.len());
+    for id in ids {
+      sorted_hashes.push(HashedId::new(id));
+    }
+    sorted_hashes.sort_by(|a, b| a.hash.cmp(&b.hash));
+    self.sorted_hashes = sorted_hashes;
+  }
+
+  pub fn get_leader_id(&self) -> &str {
+    let mut last_idx = self.sorted_hashes.len() - 1;
+    while last_idx > 0 && !self.sorted_hashes[last_idx].is_up {
+      last_idx = last_idx - 1;
+    }
+    &self.sorted_hashes[last_idx].id
+  }
+
+  pub fn get_mut_nodes(self: &mut LogRendezvousHash) -> &mut Vec<HashedId> {
+    &mut self.sorted_hashes
+  }
+
+  // Runs a binary search for the record whose hash is closest to the key hash without
+  // going over. If none are found, the *last* record in the list is returned as it wraps around.
+  pub fn _get_id_for_key(&self, key: &str) -> &str {
+    let mut key_hasher = XxHash64::with_seed(0xa1a2);
+    key_hasher.write(key.as_bytes());
+    let key_hash = key_hasher.finish();
+    let idx = match self
+      .sorted_hashes
+      .binary_search_by(|a| a.hash.cmp(&key_hash))
+    {
+      Ok(res) => res,
+      // All were too large, implies last (which wraps around) owns it
+      Err(_) => self.sorted_hashes.len() - 1,
+    };
+    &self.sorted_hashes[idx].id
+  }
+}
+
+pub struct ControlPort {
+  lrh: LogRendezvousHash,
+  client: Client<HttpsConnector<HttpConnector>>,
+  // TODO: Once the crazy type info of the server can be figured out, we can attach it to this
+  // struct and then make it possible to wind down the control port server
+  // server: &'a dyn Service<std::convert::Infallible>,
+}
+
+async fn control_port(req: Request<Body>) -> Result<Response<Body>, Infallible> {
```

Comment:
maybe `health` is a better name now?

Reply:
So I split the cluster routing/health task in two. The very next PR would add the new `/ping` endpoint so I'd prefer to keep it named this way?

Reply:
sounds good!